// @cosmwasm/ts-codegen/plugins/react-query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactQueryPlugin = void 0;
const w = __importStar(require("@cosmwasm/ts-codegen-ast"));
const ts_codegen_ast_1 = require("@cosmwasm/ts-codegen-ast");
const case_1 = require("case");
const utils_1 = require("../utils");
const plugin_base_1 = require("./plugin-base");
class ReactQueryPlugin extends plugin_base_1.BuilderPluginBase {
    initContext(contract, options) {
        return new ts_codegen_ast_1.RenderContext(contract, options, this.builder?.builderContext);
    }
    async doRender(name, context) {
        const options = this.options.reactQuery;
        const { enabled } = options;
        if (!enabled) {
            return;
        }
        const { schemas } = context.contract;
        const localname = (0, case_1.pascal)(`${name}`) + '.react-query.ts';
        const ContractFile = (0, case_1.pascal)(`${name}`) + '.client';
        const TypesFile = (0, case_1.pascal)(`${name}`) + '.types';
        const QueryMsg = (0, utils_1.findQueryMsg)(schemas);
        const ExecuteMsg = (0, utils_1.findExecuteMsg)(schemas);
        const typeHash = await (0, utils_1.findAndParseTypes)(schemas);
        const ExecuteClient = (0, case_1.pascal)(`${name}Client`);
        const QueryClient = (0, case_1.pascal)(`${name}QueryClient`);
        const body = [];
        const clientImports = [];
        QueryMsg && clientImports.push(QueryClient);
        // check that there are commands within the exec msg
        const shouldGenerateMutationHooks = ExecuteMsg &&
            options?.version === 'v4' &&
            options?.mutations &&
            (0, ts_codegen_ast_1.getMessageProperties)(ExecuteMsg).length > 0;
        if (shouldGenerateMutationHooks) {
            clientImports.push(ExecuteClient);
        }
        // general contract imports
        body.push(w.importStmt(Object.keys(typeHash), `./${TypesFile}`));
        // client imports
        body.push(w.importStmt(clientImports, `./${ContractFile}`));
        // query messages
        if (QueryMsg) {
            [].push.apply(body, w.createReactQueryHooks({
                context,
                queryMsg: QueryMsg,
                contractName: name,
                QueryClient
            }));
        }
        if (shouldGenerateMutationHooks) {
            [].push.apply(body, w.createReactQueryMutationHooks({
                context,
                execMsg: ExecuteMsg,
                contractName: name,
                ExecuteClient
            }));
        }
        if (typeHash.hasOwnProperty('Coin')) {
            // @ts-ignore
            delete context.utils.Coin;
        }
        return [
            {
                type: 'react-query',
                localname,
                body
            }
        ];
    }
}
exports.ReactQueryPlugin = ReactQueryPlugin;
