// @cosmwasm/ts-codegen-ast/utils/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROVIDER_TYPES = exports.FIXED_EXECUTE_PARAMS = exports.OPTIONAL_MEMO_PARAM = exports.OPTIONAL_FEE_PARAM = exports.OPTIONAL_FUNDS_PARAM = void 0;
const t = __importStar(require("@babel/types"));
const babel_1 = require("./babel");
exports.OPTIONAL_FUNDS_PARAM = (0, babel_1.identifier)('_funds', t.tsTypeAnnotation(t.tsArrayType(t.tsTypeReference(t.identifier('Coin')))), true);
exports.OPTIONAL_FEE_PARAM = (0, babel_1.identifier)('fee', t.tsTypeAnnotation(t.tsUnionType([
    t.tsNumberKeyword(),
    t.tsTypeReference(t.identifier('StdFee')),
    t.tsLiteralType(t.stringLiteral('auto'))
])), true);
exports.OPTIONAL_MEMO_PARAM = (0, babel_1.identifier)('memo', t.tsTypeAnnotation(t.tsStringKeyword()), true);
exports.FIXED_EXECUTE_PARAMS = [
    exports.OPTIONAL_FEE_PARAM,
    exports.OPTIONAL_MEMO_PARAM,
    exports.OPTIONAL_FUNDS_PARAM
];
exports.PROVIDER_TYPES = {
    SIGNING_CLIENT_TYPE: 'client',
    QUERY_CLIENT_TYPE: 'queryClient',
    MESSAGE_COMPOSER_TYPE: 'message-composer',
    PROVIDER_TYPE: 'provider'
};
